### PR TITLE
Add Josh Rule's primitive lists domain

### PR DESCRIPTION
**Pull request by theoxo**
This PR adds Josh Rule's primitive list domain to dreamegg.

- [x] Evaluate the "million monkeys theorem" by blindly writing code, check that it compiles
- [x] Write some basic tests, check that they pass
- [ ] Check that they synthesizer works with this domain
- [x] Write some tests with lists of lists as args, check that they pass
- [x] Write some tests using "fix", check that they pass
- [x] Handle non-terminating invocations of "fix"
- [ ] Execute some non-trivial hand-coded programs from Josh's thesis
- [x] Add error handling
- [x] Write some non-passing tests to check that error handling works as intended

### CODE REVIEW

**Comment by theoxo** [comment]
This is blocked waiting on #23 since branches in this domain need to be executed lazily.

**Comment by theoxo** [comment]
@mlb2251 I think we should just go ahead and merge this. I checked Josh's thesis but the example programs are all written using higher-level constructs (e.g. "repeat" or "reverse"), so it's not easy to translate those into the simple list domain given here.

File: src/domain.rs [hunk 1]
```diff
@@ -115,6 +115,13 @@ impl<D: Domain> CurriedFn<D> {
             partial_args: Vec::new(),
         }
     }
+    pub fn new_force_args(name: egg::Symbol, arity: usize, partial_args: Vec<LazyVal<D>>) -> Self {
```
Comment (mlb2251) [inline src/domain.rs:118]: can we rename this to `new_with_args`?

Comment (theoxo) [reply src/domain.rs]: Done!

**Pull request merged**